### PR TITLE
refactor(BUX-357): Separation of code for cron jobs

### PR DESCRIPTION
**Pull request by chris-4chain**
# Pull Request Checklist

- [x] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/BuxOrg/bux/blob/main/.github/CODE_STANDARDS.md)
- [x] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/BuxOrg/bux/blob/main/.github/CODE_OF_CONDUCT.md)
- [x] 🏠 I tested my changes locally.
- [x] ✅ I have provided tests for my changes.
- [x] 📝 I have used conventional commits.
- [ ] 📗 I have updated any related documentation.
- [x] 💾 PR was issued based on the Github or Jira issue.

I removed `registerTasks` methods from models, proposing single file declarations of all cron jobs in `cron_job_declarations.go`. Now, they are not tightly coupled with models. 

The file: `cron_job_definitions.go` (renamed from `tasks.go`) with several adjustments, has the logic of the tasks. 

The boilerplate, repetitive code of registering/running periodic tasks (cron jobs) was placed in `taskmanager/cron_jobs.go` file with tests.

I provided additional example (`examples/client/custom_cron`) - how to modify default cron jobs. By modify - I mean - updating periods, removing default jobs and adding custom ones.

The `CronJob` type in `taskmanager` remained as simple as possible. Of course - if needed - it can be extended with `Task` and `TaskOptions` properties. 

I tested it locally replacing the `bux` in `bux-server` to my local version in the `go.mod`. A transaction from HandCash to my wallet was detected and appeared as `confirmed`. Additionally I was observing logs in the `bux-server` to confirm whether the cronJobs execute.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -442,37 +428,6 @@ func (c *Client) Logger() zLogger.GormLoggerInterface {
 	return c.options.logger
 }
 
-// ModifyTaskPeriod will modify a cron task's duration period from the default
```
Comment (arkadiuszos4chain) [inline client.go:445]: <3

File: client_internal.go [hunk 2]
```diff
@@ -203,26 +196,8 @@ func (c *Client) runModelMigrations(models ...interface{}) (err error) {
 	return
 }
 
-// runModelRegisterTasks will run the model RegisterTasks() method for all models
```
Comment (arkadiuszos4chain) [inline client_internal.go:206]: <3

File: cron_job_definitions.go [hunk 3]
```diff
@@ -30,7 +28,7 @@ func taskCleanupDraftTransactions(ctx context.Context, logClient zLogger.GormLog
 
 	// Get the records
 	if err := getModels(
-		ctx, NewBaseModel(ModelNameEmpty, opts...).Client().Datastore(),
+		ctx, NewBaseModel(ModelNameEmpty, WithClient(client)).Client().Datastore(),
```
Comment (arkadiuszos4chain) [inline cron_job_definitions.go:31]: Please replace `NewBaseModel(ModelNameEmpty, WithClient(client)).Client().Datastore()` with `client.Datastore()`

File: model_access_keys.go [hunk 4]
```diff
@@ -189,11 +187,6 @@ func (m *AccessKey) BeforeCreating(_ context.Context) error {
 	return nil
 }
 
-// RegisterTasks will register the model specific tasks on client initialization
```
Comment (arkadiuszos4chain) [inline model_access_keys.go:192]: <3

File: models.go [hunk 5]
```diff
@@ -56,7 +56,6 @@ type ModelInterface interface {
 	New()
 	NotNew()
 	RawXpub() string
-	RegisterTasks() error
```
Comment (arkadiuszos4chain) [inline models.go:59]: 🎉

**Review by arkadiuszos4chain — CHANGES_REQUESTED** [review]
Awesome!

File: taskmanager/cron_jobs.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+type CronJob struct {
+	Handler CronJobHandler
+	Period  time.Duration
+}
+
+// CronJobs as a map prevents duplicate jobs with the same name
+type CronJobs map[string]CronJob
+
+// CronJobsInit registers and runs the cron jobs
+func (tm *Client) CronJobsInit(target interface{}, cronJobsMap CronJobs) (err error) {
```
Comment (dorzepowski) [inline taskmanager/cron_jobs.go:21]: does this `target` need to be `interface{}` (aka `any`)? can't it be replaced with some concrete type?

File: client_options.go [hunk 7]
```diff
@@ -582,6 +572,14 @@ func WithCronService(cronService taskmanager.CronService) ClientOps {
...
 }
 
+func WithCustomCronJobs(modifier func(cronJobs taskmanager.CronJobs) taskmanager.CronJobs) ClientOps {
+	return func(c *clientOptions) {
+		if c.taskManager != nil {
+			c.taskManager.cronJobs = modifier(c.taskManager.cronJobs)
+		}
+	}
+}
+
```
[the remark this answers is not in the document]
Comment (chris-4chain) [reply client_options.go:582]: It definitely can be removed if we don't want expose such utility. To discussion is - should we give ability to adjust time periods for each cronJobs.

File: cron_job_declarations.go [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+		Handler: BuxClientHandler(taskBroadcastTransactions),
+	},
+	CronJobNameSyncTransactionSync: {
+		Period:  120 * time.Second,
+		Handler: BuxClientHandler(taskSyncTransactions),
+	},
+}
+
+// utility function - converts a handler with the *Client target to a generic taskmanager.CronJobHandler
+func BuxClientHandler(handler func(ctx context.Context, client *Client) error) taskmanager.CronJobHandler {
```
[the remark this answers is not in the document]
Comment (chris-4chain) [reply cron_job_declarations.go:38]: If we remove `WithCustomCronJobs` we will be able to make it private.

File: taskmanager/cron_jobs.go (hunk 6 above)
Comment (chris-4chain) [reply taskmanager/cron_jobs.go:21]: A far as it is in separate package (`taskmanager`) - it is only a general utility. The main use case is to use it with `bux.client` which is in other package. So from `CronJobsInit` perspective it doesn't know the `bux.client` structure.

File: cron_job_definitions.go (hunk 3 above)
Comment (chris-4chain) [reply cron_job_definitions.go]: Thanks for noticing that. I was definitely over-complicated code. Done.

**Review by arkadiuszos4chain — APPROVED** [review]

**Pull request merged**